### PR TITLE
Migrations over WebSocket

### PR DESCRIPTION
**Pull request by SevInf**
[Internal spec](https://prisma-company.slack.com/archives/C058VM009HT/p1727364296689599?thread_ts=1727180536.749469&cid=C058VM009HT)
Close ORM-280

### CODE REVIEW

File: schema-engine/connectors/sql-schema-connector/src/flavour/postgres.rs [hunk 1]
```diff
@@ -28,9 +29,62 @@ SET enable_experimental_alter_column_type_general = true;
 
 type State = super::State<Params, (BitFlags<Circumstances>, Connection)>;
 
+#[derive(Debug, Clone)]
+struct MigratePostgresUrl(PostgresUrl);
+
+static MIGRATE_WS_BASE_URL: Lazy<Cow<'static, str>> = Lazy::new(|| {
+    std::env::var("PRISMA_SCHEMA_ENGINE_WS_BASE_URL")
```
Comment (pcan) [inline schema-engine/connectors/sql-schema-connector/src/flavour/postgres.rs:36]: FYI: the production value for the **full** ws url is `wss://migrations.prisma-data.net/websocket`. 

Refs:
- https://github.com/prisma/pdp-cloudflare/blob/migrate-ws-endpoint/services/migrations/wrangler.toml#L38
- https://github.com/prisma/pdp-cloudflare/blob/migrate-ws-endpoint/services/migrations/src/worker.ts#L57

File: quaint/src/connector/postgres/native/mod.rs [hunk 2]
```diff
@@ -292,6 +296,21 @@ impl PostgreSql {
...
     }
 
+    /// Create a new websocket connection to managed database
+    pub async fn new_with_websocket(url: PostgresWebSocketUrl) -> crate::Result<Self> {
+        let client = connect_via_websocket(url).await?;
+
+        Ok(Self {
+            client: PostgresClient(client),
+            socket_timeout: None,
+            pg_bouncer: false,
```
Comment (jkomyno) [inline quaint/src/connector/postgres/native/mod.rs:306]: Should `pg_bouncer` be always set to false, or should it be dynamically set depending on the query parameters in the connection URL?

File: quaint/src/connector/postgres/native/mod.rs [hunk 3]
```diff
@@ -292,6 +296,21 @@ impl PostgreSql {
...
+            statement_cache: Mutex::new(StatementCache::new(0)),
+            is_healthy: AtomicBool::new(true),
+            is_cockroachdb: false,
```
Comment (jkomyno) [inline quaint/src/connector/postgres/native/mod.rs:309]: No more `let is_cockroachdb = conn.parameter("crdb_version").is_some();`, then?

File: quaint/src/connector/postgres/native/websocket.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    let config = Config::from_str(connection_params)?;
+    let ws_byte_stream = WsStream::new(ws_stream);
+
+    let tls = TlsConnector::new(native_tls::TlsConnector::new()?, "TODO");
+    let (client, connection) = config.connect_raw(ws_byte_stream, tls).await?;
+    tokio::spawn(connection.map(|r| match r {
+        Ok(_) => (),
+        Err(e) => {
+            tracing::error!("Error in PostgreSQL connection: {:?}", e);
```
Comment (jkomyno) [inline quaint/src/connector/postgres/native/websocket.rs:47]:
```suggestion
            tracing::error!("Error in PostgreSQL WebSocket connection: {:?}", e);
```

File: schema-engine/connectors/sql-schema-connector/src/flavour/postgres/connection.rs [hunk 5]
```diff
@@ -5,27 +5,30 @@ use indoc::indoc;
...
-            .await
-            .map_err(quaint_err(&url))?;
+        let quaint = match url.0 {
+            PostgresUrl::Native(ref native_url) => connector::PostgreSql::new(native_url.as_ref().clone())
+                .await
+                .map_err(quaint_err(&url))?,
+            PostgresUrl::WebSocket(ref ws_url) => connector::PostgreSql::new_with_websocket(ws_url.clone())
+                .await
+                .map_err(quaint_err(&url))?,
+        };
```
Comment (jkomyno) [inline schema-engine/connectors/sql-schema-connector/src/flavour/postgres/connection.rs:31]: Nit: can we factor the duplicated `.await` and `.map_err(quaint_err(&url))?` out of the `match` expression?

E.g.:

```suggestion
        let quaint = match url.0 {
            PostgresUrl::Native(ref native_url) => connector::PostgreSql::new(native_url.as_ref().clone())
            PostgresUrl::WebSocket(ref ws_url) => connector::PostgreSql::new_with_websocket(ws_url.clone())
        }
          .await
          .map_err(quaint_err(&url))?,
```

File: quaint/src/connector/postgres/native/websocket.rs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    };
+
+    let connection_params = header.to_str().map_err(|inner| {
+        Error::builder(ErrorKind::Native(NativeErrorKind::ConnectionError(Box::new(inner)))).build()
+    })?;
+
+    let config = Config::from_str(connection_params)?;
+    let ws_byte_stream = WsStream::new(ws_stream);
+
+    let tls = TlsConnector::new(native_tls::TlsConnector::new()?, "TODO");
```
Comment (jkomyno) [inline quaint/src/connector/postgres/native/websocket.rs:42]: Leftover `"TODO"` string?

File: quaint/src/connector/postgres/native/mod.rs (hunk 2 above)
Comment (SevInf) [reply quaint/src/connector/postgres/native/mod.rs:306]: DBs that will be connecting to are guaranteed to not have pgbouncer in front of them

File: quaint/src/connector/postgres/native/mod.rs (hunk 3 above)
Comment (SevInf) [reply quaint/src/connector/postgres/native/mod.rs:309]: This guaranteed to connect only to postgres instances

File: schema-engine/connectors/sql-schema-connector/src/flavour/postgres/connection.rs (hunk 5 above)
Comment (SevInf) [reply schema-engine/connectors/sql-schema-connector/src/flavour/postgres/connection.rs:31]: Not the `await` part (different match arms return different futures), but for `map_err` we can

**Review by jkomyno — APPROVED** [review]
Approved after changes ✨

**Pull request merged**